{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using mtSet from command line\n",
    "\n",
    "Multi Trait Set test is an implementation of efficient set test algorithms for testing for associations between multiple genetic variants and multiple traits.\n",
    "mtSet can account for confounding factors such as relatedness and can be used for analysis of single traits. A detailed description of the method can be found at [1].\n",
    "\n",
    "Below we show how to use mtSet from command line using the limix scripts (`mtSet_preprocess`, `mtSet_analyze`, `mtSet_postprocess`, `mtSet_simPheno`).\n",
    "\n",
    "mtSet can also be used within python as shown in [iPython notebook tutorial](http://nbviewer.jupyter.org/github/pmbio/limix-tutorials/blob/master/mtSet/Multi_trait_set_test.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Quick Start\n",
    "\n",
    "In the following, we give a brief example on how to use mtSet. As a case study, we use a subset of the genotypes from the 1000 project [2] and simulated phenotypes.\n",
    "\n",
    "All commands can be found in `_demos/runmtSet.sh`. In the following, we give a short summary of the individual steps. A demo for running mtSet-PC can be found in `_demos/runmtSetPC.sh` and it is not showcased here.\n",
    "\n",
    "* Our software depends on [Plink](https://www.cog-genomics.org/plink2) version 1.9 (or greater) for preprocessing. Please, make sure you have it before proceeding.\n",
    "\n",
    " Install Limix (see also instructions at https://github.com/limix/limix).\n",
    "```bash\n",
    "pip install limix\n",
    "```\n",
    "\n",
    "* Download sample data from http://www.ebi.ac.uk/~casale/data.zip and unzip them\n",
    "```bash\n",
    "wget http://www.ebi.ac.uk/~casale/data.zip\n",
    "unzip data.zip\n",
    "```\n",
    "\n",
    "* Set some handy shell variables\n",
    "```bash\n",
    "BFILE=data/chrom22_subsample20_maf0.10\n",
    "CFILE=out/chrom22\n",
    "PFILE=data/pheno\n",
    "WFILE=out/windows\n",
    "NFILE=out/null\n",
    "WSIZE=30000\n",
    "STEP=15000\n",
    "RESDIR=out/results\n",
    "OUTFILE=out/final\n",
    "```\n",
    "\n",
    "* Kinship matrix estimation\n",
    "```bash\n",
    "mtSet_preprocess --compute_covariance --bfile $BFILE --cfile $CFILE \n",
    "```\n",
    "\n",
    "* Fitting the null model\n",
    "```bash\n",
    "mtSet_preprocess --fit_null --bfile $BFILE --cfile $CFILE --pfile $PFILE --nfile $NFILE\n",
    "```\n",
    "\n",
    "* Defining windows\n",
    "```bash\n",
    "mtSet_definesets --sliding_window --wfile $WFILE --bfile $BFILE --window_size $WSIZE --step $STEP --minSnps 4\n",
    "```\n",
    "\n",
    "* Analysing true genotypes\n",
    "```bash\n",
    "mtSet_analyze --bfile $BFILE --cfile $CFILE --pfile $PFILE --nfile $NFILE --wfile $WFILE --resdir $RESDIR --start_wnd 0 --end_wnd 100 --standardize --unique_variants\n",
    "```\n",
    "\n",
    "* Analysing permuted genotypes (3 permutations)\n",
    "```bash\n",
    "for i in `seq 0 2`; do\n",
    "mtSet_analyze --bfile $BFILE --cfile $CFILE --pfile $PFILE --nfile $NFILE --wfile $WFILE --resdir $RESDIR --start_wnd 0 --end_wnd 100 --perm $i --standardize --unique_variants\n",
    "done\n",
    "```\n",
    "\n",
    "* Postprocess\n",
    "```bash\n",
    "mtSet_postprocess --resdir $RESDIR --outfile $OUTFILE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing\n",
    "\n",
    "Before getting started, we have to compute the sample-to-sample genetic covariance matrix, assign the markers to windows and estimate the trait-to-trait covariance matrices on the null model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Computing the Covariance Matrix\n",
    "The covariance matrix can be pre-computed as follows:\n",
    "\n",
    "```bash\n",
    "mtSet_preprocess --compute_covariance --plink_path plink_path  --bfile bfile  --cfile cfile\n",
    "```\n",
    "\n",
    "where\n",
    "* __plink\\_path__ (default: plink) is a pointer to the [plink software](https://www.cog-genomics.org/plink2) (Version 1.9 or greater must be installed).\n",
    "  If not set, a python covariance reader is employed.\n",
    "  We strongly recommend using the plink reader for large datasets.\n",
    "* __bfile__ is the base name of of the binary bed file (__bfile__.bed, __bfile__.bim, __bfile__.fam are required).\n",
    "* __cfile__ is the base name of the output file.\n",
    "  The relatedness matrix will be written to __cfile__.cov while the identifiers of the individuals are written to the file __cfile__.cov.id.\n",
    "  The eigenvalue decomposition of the matrix is saved in the files __cfile__.cov.eval (eigenvalues) and __cfile__.cov.evec (eigenvectors).\n",
    "  If __cfile__ is not specified, the files will be exported to the current directory with the following filenames __bfile__.cov, __bfile__.cov.id, __bfile__.cov.eval, __bfile__.cov.evec."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Precomputing the Principal Components \n",
    "The principal components can be pre-computed as follows:\n",
    "\n",
    "```bash\n",
    "mtSet_preprocess --compute_PCs k --plink_path plink_path --ffile ffile  --bfile bfile\n",
    "```\n",
    "\n",
    "where\n",
    "* __k__ is the number of top principal components that are saved\n",
    "* __plink\\_path__ (default: plink) is a pointer to the [plink software](https://www.cog-genomics.org/plink2) (Version 1.9 or greater must be installed).\n",
    "  If not set, a python genotype reader is employed.\n",
    "  We strongly recommend using the plink reader for large datasets.\n",
    "* __ffile__ is the name of the fixed effects file, to which the principal components are written to.\n",
    "* __bfile__ is the base name of of the binary bed file (__bfile__.bed, __bfile__.bim, __bfile__.fam are required)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fitting the null model\n",
    "\n",
    "To efficiently apply mtSet, it is neccessary to compute the null model beforehand.\n",
    "This can be done with the following command:\n",
    "\n",
    "```bash\n",
    "mtSet_preprocess --fit_null --bfile bfile --cfile cfile --nfile nfile --pfile pfile --ffile ffile --trait_idx trait_idx\n",
    "```\n",
    "\n",
    "where\n",
    "* __bfile__ is the base name of of the binary bed file (_bfile_.bed,_bfile_.bim,_bfile_.fam are required).\n",
    "* __cfile__ is the base name of the covariance file and its eigen decomposition (__cfile__.cov, __cfile__.cov.eval and __cfile__.cov.evec).\n",
    "If __cfile__ is not set, the relatedness component is omitted from the model.\n",
    "* __nfile__ is the base name of the output file.\n",
    "The estimated parameters are saved in __nfile__.p0, the negative log likelihood ratio in __nfile__.nll0, the trait-to-trait genetic covariance matrix in __nfile__.cg0 and the trait-to-trait residual covariance matrix in __nfile__.cn0. \n",
    "* __pfile__ is the base name of the phenotype file.\n",
    "* __ffile__ is the name of the file containing the covariates. Each covariate is saved in one column\n",
    "* __trait\\_idx__ can be used to specify a subset of the phenotypes. If more than one phenotype is selected, the phenotypes have to be seperated by commas. For instance, `--trait_idx 3,4` selects the phenotypes saved in the forth and fifth column (indexing starts with zero).\n",
    "\n",
    "Notice that phenotypes are standardized prior to model fitting."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining the sets\n",
    "To apply mtSet, we need to specify a list of variant-sets we want to analyze (see below).\n",
    "This is tab-separated file having columns:\n",
    "- set identifier (*setid*)\n",
    "- chromosome (*chrom*)\n",
    "- start position (*start*)\n",
    "- end position (*end*)\n",
    "- number of variants (*nsnps*, optional)\n",
    "\n",
    "An example can be found [here](http://www.ebi.ac.uk/~casale/mtSet_demo/windows.wnd).\n",
    "\n",
    "\n",
    "We provide utility scripts to\n",
    "1. define sets using a sliding-window approach with specified window size and step;\n",
    "```bash\n",
    "mtSet_definesets --sliding_window --bfile bfile --wfile wfile --window_size window_size --step step --minSnps minSnps --maxSnps maxSnps\n",
    "```\n",
    "where\n",
    "    * __bfile__ is the base name of of the binary bed file (__bfile__.bim is required).\n",
    "    * __wfile__ is the base name of the output file.\n",
    "      This file defines the sets to analyse.\n",
    "      It is a tab-separated file having columns:\n",
    "        - set identifier (*setid*)\n",
    "        - chromosome (*chrom*)\n",
    "        - start position (*start*)\n",
    "        - end position (*end*)\n",
    "        - number of variants (*nsnps*, optional)\n",
    "    * __window\\_size__ is the size of the window (in basepairs). The default value is 30kb.\n",
    "    * __step__ is the step for the sliding-windows (in basepairs). The default value is 15kb.\n",
    "    * __minSnps__ is the minimum number of variants a region has to contain not to be excluded.\n",
    "    * __maxSnps__ is the maximum number of variants a region has to contain not to be excluded.\n",
    "\n",
    "2. filter an existing file based on the number of SNPs in the specific dataset.\n",
    "```bash\n",
    "mtSet_definesets --filter_sets --bfile bfile --wfile wfile --iwfile iwfile --minSnps minSnps --maxSnps maxSnps\n",
    "```\n",
    "where\n",
    "    * __bfile__ is the base name of of the binary bed file (__bfile__.bim is required).\n",
    "    * __wfile__ is the base name of the output file.\n",
    "      If not specified, the file is saved as __bfile__.window\\_size.wnd in the current folder.\n",
    "      This file defines the sets to analyse (see above).\n",
    "    * __iwfile__ is the input window file to filter.\n",
    "    It should have the same format as __wfile__.\n",
    "    Note that the column nsnps is optional,\n",
    "    so that only *setid*, *chrom*, *start* and *end* need to be specified.\n",
    "    * __minSnps__ is the minimum number of variants a region has to contain not to be excluded.\n",
    "    * __maxSnps__ is the maximum number of variants a region has to contain not to be excluded."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging the preprocessing steps\n",
    "\n",
    "Here, we provided the commands to execute the three preprocessing operations individually. However, it is also possible to combine all steps in a single command:\n",
    "\n",
    "```bash\n",
    "mtSet_preprocess --compute_covariance --fit_null ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype simulation\n",
    "\n",
    "Our software package also includes a command-line simulator that allows to generate phenotypes with a wide range of different genetic architectures.\n",
    "In brief, the simulator assumes a linear-additive model, considering the contribution of a randomly selected (causal) genetic region for the set component, polygenic background effects from all remaining genome-wide variants, a contribution from unmeasured factors and iid observation noise.\n",
    "For a detailed description of the simulation procedure, we refer to the Supplementary Methods. \n",
    "\n",
    "The simulator requires as input the genotypes and the relatedness component:\n",
    "\n",
    "```bash\n",
    "mtSet_simPheno --bfile bfile --cfile cfile --pfile pfile\n",
    "```\n",
    "\n",
    "where\n",
    "* __bfile__ is the name of of the binary bed file (__bfile__.bed, __bfile__.bim, __bfile__.fam are required).\n",
    "* __cfile__ is the name of the covariance matrix file (__cfile__.cov, __cfile__.cov.id are required).\n",
    "If none is specified, the covariance matrix is expected to be in the current folder, having the same filename as the bed file.\n",
    "* __pfile__ is the name of the output file (__pfile__.phe, __pfile__.region).\n",
    "The file __pfile__.phe contains the phenotypic values (each sample is saved in one row, each trait in one column).\n",
    "The file __pfile__.region contains the randomly selected causal region (chromsom, start position, end position). \n",
    "If __pfile__ is not specified, the files are saved in the current folder having an automatic generated filename containing the bed filename and the values of all simulation parameters.\n",
    "\n",
    "By changing the following parameters different genetic architectures can be simulated and, in particular, the simulation experiments of our paper can be reproduced.\n",
    "\n",
    "\n",
    "| Option        | Default       | Datatype | Explanation |\n",
    "| ------------- |:-------------:|:--------:| --------|\n",
    "| `--seed`       | 0 | int | seed for random number generator |\n",
    "| `--nTraits`    | 4 | int | number of simulated phenotypes |\n",
    "| `--windowSize` | 1.5e4 | int | size of causal region |\n",
    "| `--vTotR` | 0.05 | float |   variance explained by the causal region |\n",
    "| `--nCausalR`  | 10 | int |   number of causal variants in the region |\n",
    "| `--pCommonR` | 0.8 | float | percentage of shared causal variants |\n",
    "| `--vTotBg` | 0.4 | float  | variance explained by the polygenic background effects |\n",
    "| `--pHidden` | 0.6 | float | residual variance explained by hidden confounders (in %) |\n",
    "| `--pCommon` | 0.8 | float | background and residual signal that is shared across traits (in %) |\n",
    "| `--chrom` | None | int | specifies the chromosome of the causal region |\n",
    "| `--minPos` | None | int | specifies the min. chromosomal position of the causal region (in basepairs) |\n",
    "| `--maxPos` | None | int | specifies the max. chromosomal position of the causal region (in basepairs) |\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running analysis\n",
    "\n",
    "Once the preprocessing step has been used to obtain the genetic relatedness matrix, to fit the null model and to identify the genetic regions to be considered in the analysis, the set test can be run by the following analysis script:\n",
    "\n",
    "```bash\n",
    "mtSet_analyze --bfile bfile --cfile cfile --pfile pfile --nfile nfile --wfile wfile --ffile ffile --start_wnd start_wnd --end_wnd end_wnd --resdir rdir --trait_idx traitIdx --standardize --unique_variants\n",
    "```\n",
    "\n",
    "where\n",
    "\n",
    "- __bfile__ is the base name of of the binary bed file (__bfile__.bed, __bfile__.bim, __bfile__.fam are required).\n",
    "- __cfile__ is the base name of the covariance matrix file. The script requires the files: __cfile__.cov containing the the genetic relatedness matrix, cfile.cov.id containing the corresponding sample identifiers, __cfile__.cov.eval and __cfile__.cov.evec containing the eigenvalues and eigenvectors of the matrix. If cfile is not set, the relatedness component is omitted from the model.\n",
    "- __pfile__ is the base name of the phenotype file. The script requires the file __pfile__.phe containing the phenotype data.\n",
    "- __nfile__ is the base name of the null model file. The script requires the file __nfile__.p0 containing the optimal null model parameters. If covariates are set, it also requires the file __nfile__.f0.\n",
    "- __wfile__ is the base name of the file containing the windows to be considered in the set test. The script requires the file __wfile__.wnd.\n",
    "- __ffile__ is the name of the file containing the covariates. Each covariate is saved in one column.\n",
    "- __perm__ is the seed used when permuting the genotypes. If the option is not specified then no permutation is considered.\n",
    "- __start\\_wnd__ is the index of the start window\n",
    "- __end\\_wnd__ is the index of the end window\n",
    "- __rdir__ is the directory to which the results are exported. The results are exported in the folder rdir/perm if a permutation seed has been set, otherwise in the folder rdir/test. The output file is named __start\\_wnd\\_end\\_wnd__.res and contains results in the following format: window index, chromosome, start position, stop position, index of startposition, number of SNPs and log likelihood ratio.\n",
    "- __trait\\_idx__ can be used to specify a subset of the phenotypes. If more than one phenotype is selected, the phenotypes have to be seperated by commas. For instance __trait\\_idx__ 3,4 selects the phenotypes saved in the forth and fifth column (indexing starts with zero).\n",
    "Notice that phenotypes are standardized prior to model fitting.\n",
    "- __standardize__: if specifed, variants are standardized. We recommend to activate it.\n",
    "  It should be consistent between test and permuted data.\n",
    "- __unique_variants__: if specifed, only variants with distinct genetic profile across individuals are considered in the set. We recommend to activate it.\n",
    "  It should be consistent between test and permuted data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Postprocessing\n",
    "\n",
    "After running mtSet, the following script can be used to merge the result files and estimate the p-values (p-values are obtained by a parametric fit of the test statistics): \n",
    "\n",
    "```bash\n",
    "mtSet_postprocess --resdir resdir --outfile outfile\n",
    "```\n",
    "\n",
    "where \n",
    "* __resdir__ is a pointer to the folder containing the result files of the analysis.\n",
    "* __outfile__ is the prefix of the two output files.\n",
    "__outfile__.perm lists the test statistics (first column) and p-values (second column) of the permutated windows\n",
    "__outfile__.test contains the (index, chromosome, start position, stop position, SNP index, number of SNPs, test statistics and p-value) of each window. Each window is saved in one row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Developers\n",
    "\n",
    "- Francesco Paolo Casale (<casale@ebi.ac.uk>)\n",
    "- Barbara Rakitsch (<rakitsch@ebi.ac.uk>)\n",
    "- Danilo Horta (<horta@ebi.ac.uk>)\n",
    "- Oliver Stegle (<stegle@ebi.ac.uk>)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "[1] Casale FP, Rakitsch B, Lippert C, Stegle O. Efficient set tests for the genetic analysis of correlated traits. Nature methods. 2015 Aug 1;12(8):755-8. ([link](http://www.nature.com/nmeth/journal/v12/n8/abs/nmeth.3439.html))\n",
    "\n",
    "[2] Genomes Project, C. et al. An integrated map of genetic variation from 1,092 human genomes. Nature 491, 56-65 (2012).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "nav_menu": {},
  "toc": {
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 6,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "toc_position": {
   "height": "1211px",
   "left": "0px",
   "right": "auto",
   "top": "106px",
   "width": "215px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
