{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interaction set test (iSet)\n",
    "\n",
    "Set tests are a powerful approach for association testing between groups of genetic variants and quantitative traits.\n",
    "iSet can be applied for interaction analysis in two data designs:\n",
    "* complete design, where all individuals have been phenotyped in each context\n",
    "* stratified design, where each individual has been phenotyped in only one of the two contexts\n",
    "\n",
    "A detailed description of the method can be found at [1].\n",
    "\n",
    "In this tutorial we showcase the usage of iSet using command line a command line interface (using the limix scripts `mtSet_preprocess`, `iSet_analyze` and `iSet_postprocess`).\n",
    "\n",
    "iSet can be also used in Python as we describe in this [iPython notebook](iSet_python.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Quick Start\n",
    "\n",
    "*  Install Limix (see also instructions at https://github.com/limix/limix).\n",
    "```bash\n",
    "pip install limix\n",
    "```\n",
    "* Download sample data from http://www.ebi.ac.uk/~casale/data.zip and unzip them\n",
    "```bash\n",
    "wget http://www.ebi.ac.uk/~casale/data.zip\n",
    "unzip data.zip\n",
    "```\n",
    "* Define sets to consider in the analysis\n",
    "```bash\n",
    "BFILE=data/chrom22_subsample20_maf0.10 #bed file basename\n",
    "WFILE=data/windows #file with the sets to analyse\n",
    "mtSet_definesets --sliding_window --wfile $WFILE --bfile $BFILE --window_size $WSIZE --step $STEP --minSnps 4\n",
    "```\n",
    "* Perform set tests from window 0 to window 9 for either complete or stratified designs (see below for further information). This command can be used to run iSet on multiple cores, each analysing a set of windows (for example, 0-9, 10-19, 20-29, etc).\n",
    "    - Complete design \n",
    "     ```bash\n",
    "     PFILE=data/pheno_compl #phenotype matrix (N samples x 2)\n",
    "     FFILE=data/covs #covariates (N samples x K covariates)\n",
    "     RESDIR=results #output folder\n",
    "     iSet_analyze --bfile $BFILE --ffile $FFILE --pfile $PFILE --wfile $WFILE --resdir $RESDIR --start_wnd 0 --end_wnd 10 --n_perms 10 --unique_variants --standardize\n",
    "     ```\n",
    "\n",
    "    - Stratified design\n",
    "     ```bash\n",
    "     PFILE=data/pheno_strat #phenotype vector (N samples x 1)\n",
    "     FFILE=data/covs #covariates (N samples x N covariates)\n",
    "     IFILE=data/indicator #environment indicator vector (0/1, N samples x 1)\n",
    "     RESDIR=results #output folder\n",
    "     iSet_analyze --bfile $BFILE --ffile $FFILE --pfile $PFILE --wfile $WFILE --minSnps 4 --resdir $RESDIR --start_wnd 0 --end_wnd 10 --n_perms 10 --ifile $IFILE --unique_variants --standardize\n",
    "     ```\n",
    "* Merges all results present in RESDIR, calculate P values and exports to OUTFILE\n",
    "```bash\n",
    "OUTFILE=final\n",
    "iSet_postprocess --resdir $RESDIR --outfile $OUTFILE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining the sets\n",
    "To apply mtSet, we need to specify a list of variant-sets we want to analyze (see below).\n",
    "This is tab-separated file having columns:\n",
    "- set identifier (*setid*)\n",
    "- chromosome (*chrom*)\n",
    "- start position (*start*)\n",
    "- end position (*end*)\n",
    "- number of variants (*nsnps*, optional)\n",
    "\n",
    "An example can be found [here](http://www.ebi.ac.uk/~casale/mtSet_demo/windows.wnd).\n",
    "\n",
    "\n",
    "We provide utility scripts to\n",
    "1. define sets using a sliding-window approach with specified window size and step;\n",
    "```bash\n",
    "mtSet_definesets --sliding_window --bfile bfile --wfile wfile --window_size window_size --step step --minSnps minSnps --maxSnps maxSnps\n",
    "```\n",
    "where\n",
    "    * __bfile__ is the base name of of the binary bed file (__bfile__.bim is required).\n",
    "    * __wfile__ is the base name of the output file.\n",
    "      This file defines the sets to analyse.\n",
    "      It is a tab-separated file having columns:\n",
    "        - set identifier (*setid*)\n",
    "        - chromosome (*chrom*)\n",
    "        - start position (*start*)\n",
    "        - end position (*end*)\n",
    "        - number of variants (*nsnps*, optional)\n",
    "    * __window\\_size__ is the size of the window (in basepairs). The default value is 30kb.\n",
    "    * __step__ is the step for the sliding-windows (in basepairs). The default value is 15kb.\n",
    "    * __minSnps__ is the minimum number of variants a region has to contain not to be excluded.\n",
    "    * __maxSnps__ is the maximum number of variants a region has to contain not to be excluded.\n",
    "\n",
    "2. filter an existing file based on the number of SNPs in the specific dataset.\n",
    "```bash\n",
    "mtSet_definesets --filter_sets --bfile bfile --wfile wfile --iwfile iwfile --minSnps minSnps --maxSnps maxSnps\n",
    "```\n",
    "where\n",
    "    * __bfile__ is the base name of of the binary bed file (__bfile__.bim is required).\n",
    "    * __wfile__ is the base name of the output file.\n",
    "      If not specified, the file is saved as __bfile__.window\\_size.wnd in the current folder.\n",
    "      This file defines the sets to analyse (see above).\n",
    "    * __iwfile__ is the input window file to filter.\n",
    "    It should have the same format as __wfile__.\n",
    "    Note that the column nsnps is optional,\n",
    "    so that only *setid*, *chrom*, *start* and *end* need to be specified.\n",
    "    * __minSnps__ is the minimum number of variants a region has to contain not to be excluded.\n",
    "    * __maxSnps__ is the maximum number of variants a region has to contain not to be excluded."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running analysis\n",
    "\n",
    "The set test can be run by the following analysis script:\n",
    "\n",
    "```bash\n",
    "iSet_analyze --bfile bfile --pfile pfile --wfile wfile --ffile ffile --minSnps minSnps --start_wnd start_wnd --end_wnd end_wnd --resdir rdir --ifile $IFILE --n_perms 10 --unique_variants --standardize\n",
    "```\n",
    "\n",
    "where\n",
    "\n",
    "- __bfile__ is the base name of of the binary bed file (__bfile__.bed, __bfile__.bim, __bfile__.fam are required).\n",
    "- __pfile__ is the base name of the phenotype file. The script requires the file __pfile__.phe containing the phenotype data.\n",
    "- __wfile__ is the base name of the file containing the windows to be considered in the set test. The script requires the file __wfile__.wnd.\n",
    "- __ffile__ is the name of the file containing the covariates. Each covariate is a column in the matrix.\n",
    "- __start\\_wnd__ is the index of the start window\n",
    "- __end\\_wnd__ is the index of the end window\n",
    "- __n_perms__ number of null (sampled) test statistics (obtained thrugh permutations/parametric bootstraps). Default is 10.\n",
    "- __rdir__ is the directory to which the results are exported. The command exports files *start_wnd*_*end_wnd*.iSet.real that contains test statistics and vairance components and *start_wnd*_*end_wnd*.iSet.perm that contains null statistics\n",
    "- __ifile__ is the file path to a csv file containing an indicator (True or False) for each sample. If specified the analysis is performed for a stratified design.\n",
    "- __startwnd\\_endwnd__.res and contains results in the following format: window index, chromosome, start position, stop position, index of startposition, number of SNPs and log likelihood ratio.\n",
    "- __standardize__: if specifed, variants are standardized. We recommend to activate it.\n",
    "  It should be consistent between test and permuted data.\n",
    "- __unique_variants__: if specifed, only variants with distinct genetic profile across individuals are considered in the set. We recommend to activate it.\n",
    "  It should be consistent between test and permuted data.\n",
    "\n",
    "Note that this command can be used to run iSet on multiple cores, each analysing a set of windows (for example, 0-9, 10-19, 20-29, etc)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Postprocessing\n",
    "\n",
    "After running iSet, the following script can be used to merge the result files and estimate the p-values (p-values are obtained by a parametric fit of the test statistics): \n",
    "\n",
    "```bash\n",
    "iSet_postprocess --resdir resdir --outfile outfile --strat\n",
    "```\n",
    "\n",
    "where \n",
    "* __resdir__ is a pointer to the folder containing the result files of the analysis.\n",
    "* __outfile__ is the prefix of the two output files.\n",
    "__outfile__.perm lists the test statistics (first column) and p-values (second column) of the permutated windows\n",
    "__outfile__.test contains the (index, chromosome, start position, stop position, SNP index, number of SNPs, test statistics and p-value) of each window. Each window is saved in one row.\n",
    "* __strat__ is a boolean flag that indicates stratified design when used."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[1] Casale FP, Horta D, Rakitsch B, Stegle O. Joint genetic analysis using variant sets reveals polygenic gene-context interactions. PLoS Genetics. 2017. ([link](http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006693))"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "nav_menu": {},
  "toc": {
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 6,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "toc_position": {
   "height": "1211px",
   "left": "0px",
   "right": "auto",
   "top": "106px",
   "width": "215px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
